microprofile/cdi/src/main/java/io/helidon/microprofile/cdi/HelidonProxyServices.java

```diff
@@ -88,11 +98,23 @@ public boolean supportsClassDefining() {
                                 int len,
                                 ProtectionDomain protectionDomain) throws ClassFormatError {
 
-        if (samePackage(originalClass, className)) {
-            return defineClassSamePackage(originalClass, className, classBytes, off, len);
-        } else {
-            return wrapCl(originalClass.getClassLoader())
-                    .doDefineClass(originalClass, className, classBytes, off, len, protectionDomain);
+        // always try to define in private lookup, if fails (and not same package), try using a different classloader
+        try {
+            return defineClassPrivateLookup(originalClass, className, classBytes, off, len);
+        } catch (Exception e) {
+            if (samePackage(originalClass, className)) {
+                // when same package, we must use private lookup (as we may use package local)
+                throw e;
+            } else {
+                LOGGER.log(Level.FINEST,
+                           "Failed to create class " + className + " in same classloader. Will use a different one",
+                           e);
+
+                // other cases (except for a few edge cases, such as producer in a different package and usage
+                // of bean in the same package) we can live with a different classloader to hold the proxy class
+                return wrapCl(originalClass.getClassLoader())
+                        .doDefineClass(originalClass, className, classBytes, off, len, protectionDomain);
```

Comment:
I don't know enough about this to have a strong opinion, but let's say the private lookup fails.  Shouldn't we still endeavor to use [`MethodHandles.Lookup#defineClass(byte[])`](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#defineClass(byte[])) to define the class instead of widening the access of `ClassLoader#defineClass(byte[])`?  Maybe using a public lookup rather than a private one?

Reply:
Will try

Reply:
`PublicLookup` is not useful, as it does not have the `PACKAGE` right, so we cannot define classes using it.
I have changed the code to work as follows:
- Weld specific packages (weldx, weld.lang) use the original classloader based approach
- all other beans use private lookup

If the private lookup fails, the bean class creation fails as well with appropriate message.
There may be some edge cases not covered by this approach, though our repository does not contain any such case.